internal/common/module_mappings.bzl

```diff
@@ -51,92 +49,59 @@ def get_module_mappings(label, attrs, srcs = [], workspace_name = None, mappings
-                    # without the trailing slash will be used only when importing from the
-                    # bare module_name.
-                    mappings[mn + "/"] = mr + "/"
-                    mr = "%s/%s" % (mr, attrs.module_root.replace(".d.ts", ".js"))
-                else:
-                    # This is the type-checking module mapping. Strip the trailing .d.ts
-                    # as it doesn't belong in TypeScript's path mapping.
-                    mr = "%s/%s" % (mr, attrs.module_root.replace(".d.ts", ""))
-
+    if hasattr(ctx.rule.attr, "module_name") and ctx.rule.attr.module_name:
+        mn = ctx.rule.attr.module_name
+
+        # When building a mapping for use at runtime, we need paths to be relative to
+        # the runfiles directory. This requires the workspace_name to be prefixed on
+        # each module root.
+        mr = "/".join([p for p in [workspace_name, target.label.package] if p])
+        if hasattr(ctx.rule.attr, "module_root") and ctx.rule.attr.module_root and ctx.rule.attr.module_root != ".":
+            if ctx.rule.attr.module_root.endswith(".ts"):
                 # Validate that sources are underneath the module root.
                 # module_roots ending in .ts are a special case, they are used to
                 # restrict what's exported from a build rule, e.g. only exports from a
                 # specific index.d.ts file. For those, not every source must be under the
                 # given module root.
-
+                #
+                # .d.ts module_root means we should be able to load in two ways:
+                #   module_name -> module_path/module_root.js
+                #   module_name/foo -> module_path/foo
+                # So we add two mappings. The one with the trailing slash is longer,
+                # so the loader should prefer it for any deep imports. The mapping
+                # without the trailing slash will be used only when importing from the
+                # bare module_name.
+                mappings[mn + "/"] = mr + "/"
+                mr = "%s/%s" % (mr, ctx.rule.attr.module_root.replace(".d.ts", ".js"))
             else:
-                mr = "%s/%s" % (mr, attrs.module_root)
-                for s in srcs:
-                    if not s.short_path.startswith(mr):
-                        fail(("all sources must be under module root: %s, but found: %s" %
-                              (mr, s.short_path)))
+                mr = "%s/%s" % (mr, ctx.rule.attr.module_root)
         if mn in mappings and mappings[mn] != mr:
             fail(("duplicate module mapping at %s: %s maps to both %s and %s" %
-                  (label, mn, mappings[mn], mr)), "deps")
+                  (target.label, mn, mappings[mn], mr)), "deps")
         mappings[mn] = mr
-    _debug("Mappings at %s: %s", (label, mappings))
+    _debug(ctx.var, "Mappings at %s: %s" % (target.label, mappings))
     return mappings
 
```

Comment:
I'm scared that we don't have enough coverage for this code and something might break. I don't really understand how you're able to get rid of the distinction between the runfiles mapping and the original execroot mapping either.

I was hoping to completely delete this file since the linker has its own aspect logic to collect module mappings. I guess it's a step in that direction.

Reply:
Removing entirely is the next step :)
We weren't using the original mappings function in this repo anymore. rules_typescript repo has a copy of the file for its aspect and that is the only spot the old function is still needed.